File: sliding window/java/LongestRepeatingCharacterReplacement.java
import java.util.HashMap;

public class LongestRepeatingCharacterReplacement {

    public static void main(String[] args) {
        
        System.out.println(findLength("aabccbb", 2));
        System.out.println(findLength("abbcb", 1));
        System.out.println(findLength("abccde", 1));
        
    }

    public static int findLength(String string, int k) {

        int result = 0;
        int windowStart = 0;
        HashMap<Character, Integer> characterLength = new HashMap<>();
        int maxRepeatingCharacter = 0;

        for(int windowEnd = 0; windowEnd < string.length(); windowEnd++){

            char c = string.charAt(windowEnd);
            characterLength.put(c, characterLength.getOrDefault(c, 0) + 1);
            maxRepeatingCharacter = Math.max(maxRepeatingCharacter, characterLength.get(c));
            System.out.println(maxRepeatingCharacter);

            if(( windowEnd - windowStart + 1 - maxRepeatingCharacter) > k ) {
                char f = string.charAt(windowStart);
                characterLength.put(f, characterLength.get(f) - 1);

                windowStart++;

            }

            result = Math.max(result, windowEnd - windowStart + 1);
        }

        return result;
        
    }
}

File: sliding window/typescript/LongestRepeatingCharacterReplacement.ts
function characterReplacement(s: string, k: number): number {

    let windowStart = 0;
    let result = 0;
    let characterMap = new Map<string, number>();
    let mostRepeated = 0;

    for (let i = 0; i < s.length; i++) {
        let char = s[i];
        let charNum = (characterMap.get(char) ?? 0) + 1;
        characterMap.set(char, charNum);
        mostRepeated = Math.max(mostRepeated, charNum);

        if ((i - windowStart + 1 - mostRepeated) > k) {
            let charS = s[windowStart];
            let charSNum = (characterMap.get(charS) ?? 1) - 1;
            characterMap.set(charS, charSNum);

            windowStart++;
        }

        result = Math.max(result, i - windowStart + 1);

    }

    return result;

};

console.log(characterReplacement("aabccbb", 2));
console.log(characterReplacement("abbcb", 1));
console.log(characterReplacement("abccde", 1));